File: python/sieve/sieve.py
import math

def sieve(max_num):
    if max_num == 0 or max_num == 1:
        return []
    candidates = [True] * (max_num)
    candidates[0]  = False
    for i in range(2, int(math.sqrt(max_num)+1)):
        if candidates[i-1] == True:
            for j in range(i*i, max_num+1, i):
                candidates[j-1] = False
    return [i+1 for i, c in enumerate(candidates) if c == True]

File: cpp/sieve/sieve.cpp
#include "sieve.h"
#include <vector>
#include <math.h>

std::vector<int> sieve::primes(int max_num)
{    
    if (max_num == 0 or max_num == 1) 
    {
        return {};
    }
    std::vector<bool> candidates(max_num + 1, true);
    
    for(int i = 2; i < sqrt(max_num)+1; i++) 
    {
        if (candidates[i] == true)
        {
            for(int j = i * i; j < max_num + 1; j+= i)
            {
                candidates[j] = false;
            }
        }
    }
    std::vector<int> primes;
    for (int i = 2; i <= max_num; i++)
    {
        if (candidates[i]) 
        {
            primes.push_back(i);
        }
    }
    return primes;
}
